dali/operators/util/randomizer.cuh

```diff
@@ -37,6 +38,21 @@ struct curand_states {
     return states_[idx];
   }
 
+  DALI_HOST inline size_t length() const {
+    return len_;
+  }
+
+  DALI_HOST inline std::shared_ptr<curandState> get_states(AccessOrder order) const {
+    auto states = mm::alloc_raw_shared<curandState, mm::memory_kind::device>(len_);
+    cudaMemcpyAsync(states.get(), states_, sizeof(curandState) * len_,
+                    cudaMemcpyDeviceToDevice, order.stream());
```

Comment:
> I also have more general question regarding synchronous operations. Are there any plans to make the checkpointing to run on some stream? Now we do the synchronous copies in few places. (I don't see this as a bad thing, the change is simpler thanks to that, but it may have some impact in cases where the checkpoints are created more often).

@klecki `.checkpoint()` returns a serialized checkpoint to the user as a normal python string, so by design it requires synchronization with the CPU. Hypothetically we could make it asynchronous by returning a promise of a checkpoint or letting the user provide a callback, but I guess it's a bit of overkill and don't plan to do it in the foreseeable future. So unless @stiepan foresees some use case for that I'd keep serialization & checkpoint restoring synchronous.

The only place where I definetely want the copy to be asynchronous is in `get_states`, which will be called after every iteration. This one will be asynchronous, right?

Reply:
I'd refrain from introducing asynchronous API. Or adding an asynchronous "CPU" stage after the gpu copy for serialization. 

There's one thing that the initial desing misses though and we could revisit it. It seems (maybe that's what @klecki has in mind too) that now when user actually wants a checkpoint, the usage of orders will trigger synchronization with the GPU stage stream. Which means we will wait for all the prefetched GPU batches to complete when, in fact, we only need the back of the GPU queue to be ready. If user wants, for some reason, to make the checkpoints every itration all the time, it may kill GPU prefetching.

Reply:
After offline discussion: I'll optimize the waiting to be more granular in a follow-up PR